src/programs/city/defense.js

```diff
@@ -131,6 +153,20 @@ class CityDefense extends kernel.process {
     const hazardTypes = [ATTACK, RANGED_ATTACK, HEAL, WORK, CLAIM]
     return _.some(hostile.body, b => _.include(hazardTypes, b.type))
   }
+
+  calculateWithFallOff (optimalValue, distance) {
+    let factor = 0.0
+
+    if (distance < TOWER_OPTIMAL_RANGE) {
```

Comment:
Can also be written as:
```javascript
let effect = optimalValue;
if(distance > TOWER_OPTIMAL_RANGE) {
    if(distance> TOWER_FALLOFF_RANGE) {
        distance= TOWER_FALLOFF_RANGE;
    }
    effect -= effect * TOWER_FALLOFF * (distance - TOWER_OPTIMAL_RANGE) / (TOWER_FALLOFF_RANGE - TOWER_OPTIMAL_RANGE);
}
return Math.floor(effect);
```
Related to the calculation code in the engine
https://github.com/screeps/engine/blob/master/src/processor/intents/towers/attack.js#L34